# Analyzing Used Car Listings on eBay Kleinanzeigen

In this project, we'll be working with a dataset of used cars from _eBay Kleinanzeigen_, a classifieds section of the German eBay website. The [original dataset](https://www.kaggle.com/orgesleka/used-cars-database/data) was scraped annd uploaded to Kaggle. The dataset that we'll work with has been modified by DataQuest for learning purposes. DataQuest's modifications include:

* Sampled 50,000 data points from the full dataset, to ensure code runs quickly in this hosted environment
* Dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

The data dictionary provided with the data is as follows:

* `dateCrawled` &mdash; When this ad was first crawled. All field-values are taken from this date
* `name` &mdash; Name of the car
* `seller` &mdash; Whether the seller is private or a dealer
* `offerType` &mdash; The type of listing
* `price` &mdash; The price on the ad to sell the car
* `abtest` &mdash; Whether the listing is included in an A/B test
* `vehicleType` &mdash; The vehicle type
* `yearOfRegistration` &mdash; The year in which the car was first registered
* `gearbox` &mdash; The transmission type
* `powerPS` &mdash; The power of the car in PS
* `model` &mdash; The car model name
* `kilometer` &mdash; How many kilometers the car has driven
* `monthOfRegistration` &mdash; The month in which the car was first registered
* `fuelType` &mdash; What type of fuel the car uses
* `brand` &mdash; The brand of the car
* `notRepairedDamage` &mdash; If the car has a damage which is not yet repaired
* `dateCreated` &mdash; The date on which the eBay listing was created
* `nrOfPictures` &mdash; The number of pictures in the ad
* `postalCode` &mdash; The postal code for the location of the vehicle
* `lastSeenOnline` &mdash; When the crawler saw this ad last online

The goal for this project is to clean the data and annalyze the included used car listings.

Let's begin.

## Importing and Observing the Data

We'll use the pandas and NumPy library to read and manipulate the `autos.csv` dataset.

In [1]:
import numpy as np
import pandas as pd

In [2]:
autos = pd.read_csv('autos_dirty.csv', encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfle...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfle...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

By observing the dataset's information and first few rows, we can see that:

* The dataset has 20 columns
* The column names use camelcase instead of Python's preferred snakecase
* The `vehicleType`, `gearbox`, `model`, `fuelType`, `notRepairedDamage` columns have missing values
* The `date` and `odometer` columns are `object` types, but could be more useful if converted to `int` or `float` type

## Cleaning the Column Names
Let's change the column names from camelcase to snakecase.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
new_columns = ['date_crawled',
               'name',
               'seller',
               'offer_type',
               'price',
               'abtest',
               'vehicle_type',
               'registration_year',
               'gearbox',
               'power_PS',
               'model',
               'odometer',
               'registration_month',
               'fuel_type',
               'brand',
               'unrepaired_damage',
               'ad_created',
               'ad_pictures',
               'postal_code',
               'last_seen']

autos.columns = new_columns

In [7]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfle...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  ad_pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Changing the column names from camelcase to snakecase makes the data easier to work with. Some of the column names were altered to more accurately describe the columns.

## Initial Data Exploration and Cleaning

Now, let's do some basic data exploratoin to determine what other cleaning tasks need to be done. When we first took a look at the dataset's information, we noticed that the `price` and `odometer` columns contained numeric data stored as text, so we'll clean and convert those into `int` or `float` type. We will also check if any columns contain data where all or almost all values are the same&mdash;these columns can be dropped as they don't have useful information for analysis. 

In [8]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-05 16:57:05  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  ad_pictures   postal_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

By observing the `top` and `freq` rows, we found two columns where almost all values are the same:
* `seller`
* `offer_type`

Furthermore, it looks like all values in the `ad_pictures` column are `0`.

In [9]:
autos['ad_pictures'].value_counts()

0    50000
Name: ad_pictures, dtype: int64

We'll remove these three columns.

In [10]:
autos = autos.drop(['seller', 'offer_type', 'ad_pictures'], axis=1)

Now, let's take a closer look at the `price` and `odometer` columns.

In [11]:
autos[['price', 'odometer']]

price   odometer
0       $5,000  150,000km
1       $8,500  150,000km
2       $8,990   70,000km
3       $4,350   70,000km
4       $1,350  150,000km
...        ...        ...
49995  $24,900  100,000km
49996   $1,980  150,000km
49997  $13,200    5,000km
49998  $22,900   40,000km
49999   $1,250  150,000km

[50000 rows x 2 columns]

In the `price` column:
* we can make the assumption that all prices are given in complete dollar amounts, i.e. none of the prices contain cents. So, we can just convert the values to `int`
* to convert the values to `int` dtype, we'll have to remove the "$" and the "," first

In [12]:
autos['price'] = (autos['price']
                        .str.replace('$', '')
                        .str.replace(',', '')
                        .astype(int))

In [13]:
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In the `odometer` column:
* we can assume that all values end with "km" and will most likely contain a ","
* we'll remove those and convert the values to `int`

In [14]:
autos['odometer'] = (autos['odometer']
                           .str.replace(',', '')
                           .str.replace('km', '')
                           .astype(int))

In [15]:
autos['odometer'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64

Since the data in the `odometer` column no longer displays the distance units, we'll rename the column to display that information.

In [16]:
autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)

## Exploring `price` and `odometer_km`

Now that we've removed the data that were obviously useless and converted the `price` and `odometer` columns into a more useful dtype, we'll explore the data further to see if there is any data that doesn't look right. We'll start with the `price` and `odometer_km` columns.

In [17]:
print('unique values:', autos['price'].unique().shape[0], '\n')
print('describe:')
print(autos['price'].describe(), '\n')
print('value_counts:')
print(autos['price'].value_counts().head(24))

unique values: 2357 

describe:
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64 

value_counts:
0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
1100     376
1300     371
3000     365
550      356
Name: price, dtype: int64


From the 50,000 row sample, there are 2,357 unique price values.

The `describe()` output shows that there are cars listed with a price of as little as \\$0 and as high as \\$100,000,000.

We can see from the `value_counts` that there are 1,421 cars listed with a $0 price. Given that these only account for about 2.8\% of the sample, we can consider removing those listings.

$100,000,000 seems a bit too high for a car listed on eBay, so let's take a look at the highest listed prices.

In [18]:
print((autos['price']
             .value_counts()
             .sort_index(ascending=False)
             .head(24)))
print('    ...')
print((autos['price']
             .value_counts()
             .sort_index(ascending=False)
             .tail(24)))

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
Name: price, dtype: int64
    ...
49       4
47       1
45       4
40       6
35       1
30       7
29       1
25       5
20       4
18       1
17       3
15       2
14       1
13       2
12       3
11       2
10       7
9        1
8        1
5        2
3        1
2        3
1      156
0     1421
Name: price, dtype: int64


We can see that there are actually 14 cars listed around or above \\$1,000,000. We also notice that after \\$350,000 the price jumps to almost \\$1,000,000, indicating that listings above \\$350,000 are the outliers of this sample. So, we'll remove those listings.

On the lower end of the prices, there are a number of cars listed at less than \\$100, which seems like an unreasonable price for a car. But, knowing that eBay is an auction site, there could be legitimate listings of cars with current bids at those prices. So, we'll keep those low-price cars.

Recap: We're getting rid of car listings with prices of \\$0 and above \\$350,000, and keeping everything in between.

In [19]:
autos = autos[autos['price'].between(1, 351000)]
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

Now, let's explore the `odometer_km` column.

In [20]:
print('unique values:', (autos['odometer_km']
                               .unique()
                               .shape[0]),
     '\n')
print('describe:')
print(autos['odometer_km'].describe(), '\n')
print('value_counts:')
print(autos['odometer_km'].value_counts())

unique values: 13 

describe:
count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64 

value_counts:
150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
5000        836
40000       815
30000       780
20000       762
10000       253
Name: odometer_km, dtype: int64


We can see that a majority of cars listed on the website are high-mileage cars. This could indicate that most of the cars were highly used or old. We can determine this by observing the other columns in the `autos` dataset.

## Exploring the Date Columns

Let's move on to the date columns and understand the date range that the data covers. The five columns that should represent date values include:
* `date_crawled`
* `registration_year`
* `registration_month`
* `ad_created`
* `last_seen`

In [21]:
autos[['date_crawled',
       'registration_year',
       'registration_month',
       'ad_created',
       'last_seen']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48565 entries, 0 to 49999
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        48565 non-null  object
 1   registration_year   48565 non-null  int64 
 2   registration_month  48565 non-null  int64 
 3   ad_created          48565 non-null  object
 4   last_seen           48565 non-null  object
dtypes: int64(2), object(3)
memory usage: 2.2+ MB


We can see that the `date_crawled`, `ad_created`, and `last_seen` columns are identified as string values. To be able to understand these columns quantitatively, we need to convert the data into a numerical representation.

In [22]:
autos[['date_crawled', 'ad_created', 'last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

Each of these columns has identical formatting that follows:

YYYY-MM-DD 24-Hour:Minute:Second

The actual date has a greater significance than the time during the day, so we'll disregard the time.

To understand the data in these columns quantitatively, we'll represent the dates as percentages.

In [23]:
print('date_crawled:')
print((autos['date_crawled']
             .str[:10]
             .value_counts(normalize=True, dropna=False)
             .sort_index()))
print('\n')
print((autos['date_crawled']
             .str[:10]
             .value_counts(normalize=True, dropna=False)
             .sort_values()))

date_crawled:
2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64


2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-0

We see that all of the car listings in our sample were visited in March and early April of 2016, and that the number of visits per day are relatively uniform.

In [24]:
print('ad_created:')
print((autos['ad_created']
             .str[:10]
             .value_counts(normalize=True, dropna=False)
             .sort_index()), '\n')
print((autos['ad_created']
             .str[:10]
             .value_counts(normalize=True, dropna=False)
             .sort_values()))

ad_created:
2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64 

2016-02-07    0.000021
2016-02-01    0.000021
2016-01-07    0.000021
2016-02-17    0.000021
2016-02-08    0.000021
                ...   
2016-03-12    0.036755
2016-04-04    0.036858
2016-03-21    0.037579
2016-03-20    0.037949
2016-04-03    0.038855
Name: ad_created, Length: 76, dtype: float64


There are 76 unique dates when an ad was created. The oldest ad was created on `2015-06-11` and the most recent on `2016-04-07`.

In [25]:
print('last_seen:')
print((autos['last_seen']
             .str[:10]
             .value_counts(normalize=True, dropna=False)
             .sort_index()))

last_seen:
2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64


The date last seen refers to when the crawler last saw the ad online. This also gives us the date at which the listing was removed, most likely because the car was sold.

In [26]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

We can see that the data suggests that the average registration year of a car is 2004. However, if we look at the minimum and maximum values, we can see that the oldest car was registered in the year 1000, and the newest car was registered in the year 9999. Cars accessible to the masses weren't invented until the early 1900s, and the year 9999 is about 8,000 years into the future. Clearly, we have some errors in our data.

## Handling Incorrect Registration Year Data

#### Registrations from the future:
These are fairly simple to deal with. A car can't be first registered after the listing was seen. So, any car with a registration year beyond 2016 is inaccurate and can be removed from the dataset.

#### Registrations from the past:
We know that the first cars available to the public were released in first decade of the 1900s. We don't know for sure when the earliest possible registration year could be, but we could estimate it to the year 1900.

Now, we'll count the number of listings with cars that fall outside the 1900 - 2016 range and see if it's safe to remove those rows entirely.

In [27]:
outside_range = (~autos['registration_year']
                       .between(1900, 2016)).sum()
data_shape = autos.shape[0]
print('registration years outside 1900-2016:', outside_range)
print('percent outside range: {:.2f}%'.format(100*(outside_range/data_shape)))

registration years outside 1900-2016: 1884
percent outside range: 3.88%


Since the registration years outside of our proposed range only accounts for 3.88\% of our data, it would be safe to remove those rows.

In [28]:
autos = autos[autos['registration_year'].between(1900, 2016)]
(autos['registration_year']
       .value_counts(normalize=True)
       .sort_values(ascending=False)
       .head(24))

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
2008    0.047450
2009    0.044665
1997    0.041794
2011    0.034768
2010    0.034040
1996    0.029412
2012    0.028063
1995    0.026285
2016    0.026135
2013    0.017202
2014    0.014203
1994    0.013474
1993    0.009104
2015    0.008397
Name: registration_year, dtype: float64

We can see that most of the cars on eBay Kleinanzeigen were registered between 1995 and 2016.

## Exploring Vehicles by Brand

In [29]:
autos['brand'].value_counts(normalize=True)*100

volkswagen        21.126368
bmw               11.004477
opel              10.758124
mercedes_benz      9.646323
audi               8.656627
ford               6.989996
renault            4.714980
peugeot            2.984083
fiat               2.564212
seat               1.827296
skoda              1.640925
nissan             1.527388
mazda              1.518819
smart              1.415994
citroen            1.400998
toyota             1.270324
hyundai            1.002549
sonstige_autos     0.981127
volvo              0.914719
mini               0.876159
mitsubishi         0.822604
honda              0.784045
kia                0.706926
alfa_romeo         0.664082
porsche            0.612669
suzuki             0.593389
chevrolet          0.569825
chrysler           0.351321
dacia              0.263490
daihatsu           0.250637
jeep               0.227073
subaru             0.214220
land_rover         0.209936
saab               0.164949
jaguar             0.156381
daewoo             0

We see that the top five car brands (Volkswagen, BMW, Opel, Mercedes Benz, and Audi) represent about 60\% of the cars listed on the website. From the list above, we can see that many car brands have a minimal presence in the car listings. So, we'll only observe cars with a listing percentage above 5\%.

In [30]:
brand_percentage = autos['brand'].value_counts(normalize=True)
brands_above_5 = brand_percentage[brand_percentage > 0.05].index
print(brands_above_5)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [31]:
avg_brand_prices = {}

for brand in brands_above_5:
    listings = autos[autos['brand'] == brand]
    avg_price = listings['price'].mean()
    avg_brand_prices[brand] = int(avg_price)
    
abp_series = pd.Series(avg_brand_prices)
abp_series.sort_values(ascending=False)

audi             9336
mercedes_benz    8628
bmw              8332
volkswagen       5402
ford             3749
opel             2975
dtype: int64

Average Price Per Car by Brand (Top 6 Brands):
* Audi: \\$9,336
* Mercedes Benz: \\$8,628
* BMW: \\$8,332
* Volkswagen: \\$5,402
* Ford: \\$3,749
* Opel: \\$2,975

We see that Audi, Mercedes Benz, and BMW are the most expensive, ranging on average between \\$8,000 and \\$10,000. Ford and Opel are the cheapest, and Volkswagen is in between.

Given that the webiste caters towards customers located in Germany, it makes sense that 5 out of the 6 most popular cars are made by German companies. 

Audi, Mercedes Benz, and BMW are known to produce high quality luxury cars, which might explain the higher average prices and the higher demand in the market. 

Opel cars are local to Germany and are offered at a considerably cheaper price. These two factors together make the car an extremely accessible car to a wide range of German consumers, explaining the cars popularity in the eBay Kleinanzeigen market.

Volkswagen cars on this market are offered at somewhat of a mid-ranged price, while still providing the qualities of a good German-engineered car. This gives consumers the idea that they can get the best value for cost on a car, which may be why Volkswagen cars account for about 21\% of this market.

Finally, Ford offers affordable cars with decent quality, which explains the low average price for their cars on this market. On top of the price and quality, Ford has turned itself into one of the most iconic and recognizable car brands in the world, justifying its popularity in the German market.

## Exploring Vehicles by Mileage

For the top 6 brands that we determined above, let's find out the average mileage for those cars and if there's any visible link to the average price.

In [32]:
avg_brand_mileage = {}

for brand in brands_above_5:
    listings = autos[autos['brand'] == brand]
    avg_km = listings['odometer_km'].mean()
    avg_brand_mileage[brand] = int(avg_km)
    
abm_series = pd.Series(avg_brand_mileage)
abm_series = abm_series.sort_index()
abp_series = abp_series.sort_index()

brand_price_km = pd.DataFrame(abp_series, columns=['avg_price'])
brand_price_km['avg_odometer_km'] = abm_series

brand_price_km.head()

avg_price  avg_odometer_km
audi                9336           129157
bmw                 8332           132572
ford                3749           124266
mercedes_benz       8628           130788
opel                2975           129310

We can see that the average mileage for cars in the top six brands are all fairly similar. This indicates that the price of these cars on the eBay Kleinanzeigen market are not related to the cars' mileage.